### ISSUE TITLE
The operator interface may be too conducive to leaking a Capability / livelock

### ISSUE THREAD

**Issue by utaal** [frankmcsherry/timely-dataflow#121]
`Capability` can be stashed in data structures in the operator logic. This does not require `.clone()`ing them:

```rust
let stash = HashMap::new();

while let Some(|cap, data|) = input.next() {
  // mem::swap data
  stash.insert(cap, data);
}
```

If these capabilities aren't `drop`ped, the operator will never release them to the rest of the system and cause a livelock. We may want to require the user to explicitly express the intent to retain a capability by only handing out a reference to the cap, and adding a `fn retain()` that returns an owned copy.

**Comment by JosephWagner** [comment]
I've totally done this!

**Comment by frankmcsherry** [comment]
Just to be clear on this (not that anyone is necessarily confused), stashing capabilities is 100% an intended feature. But this "error" has shown up a bunch that something about it seems like a footgun, and we are sniffing around for equally idiomatic, but less risky ways to let folks have access to timely's synchronization.

Perhaps one analogy is with `Mutex`, which does not attempt to lock the mutex and hand you the lock each time you access a struct containing a `Mutex`. It could, and you learn that stashing the lock is pretty dangerous, but really no need to do that in the first place.

I currently really like the idea of handing back a `&Capability<T>` much like how the data are also only a reference (though a `&mut` reference). In addition to preventing casual stashing without a `clone()`, we don't need to perform any capability computation to hand out a `&Capability<T>`, deferring the work to the `clone()` call. This should be an easy-to-implement change, but it is also (intentionally) massively breaking.

**Comment by frankmcsherry** [comment]
Should be fixed by #132.
